zeth_contracts/contracts/BaseMixer.sol

```diff
@@ -512,8 +462,62 @@ abstract contract MixerBase is BaseMerkleTree, ERC223ReceivingContract
         }
     }
 
-    function addMerkleRoot(bytes32 root) internal
+    function _addMerkleRoot(bytes32 root) internal
     {
         _roots[root] = true;
     }
+
+    // ======================================================================
+    // Reminder: Remember that the primary inputs are ordered as follows:
+    //
+    //   [Root, CommitmentS, NullifierS, h_sig, h_iS, Residual Element(s)]
+    //
+    // ie, below is the index mapping of the primary input elements on the
+    // protoboard:
+    //
+    //   <Merkle Root>               0
+    //   <Commitment[0]>             1
+    //   ...
+    //   <Commitment[_JSOUT - 1]>     _JSOUT
+    //   <Nullifier[0]>              _JSOUT + 1
+    //   ...
+    //   <Nullifier[_JSIN]>           _JSOUT + _JSIN
+    //   <h_sig>                     _JSOUT + _JSIN + 1
+    //   <Message Auth Tag[0]>       _JSOUT + _JSIN + 2
+    //   ...
+    //   <Message Auth Tag[_JSIN]>    _JSOUT + 2*_JSIN + 1
+    //   <Residual Field Elements>   _JSOUT + 2*_JSIN + 2
+    //
+    // The Residual field elements are structured as follows:
+    //
+    //   255                                         128         64           0
+    //   |<empty>|<h_sig>|<nullifiers>|<msg_auth_tags>|<v_pub_in>)|<v_pub_out>|
+    //
+    // where each entry entry after public output and input holds the
+    // (curve-specific) number residual bits for the corresponding 256 bit
+    // value.
+    // ======================================================================
+
+    /// Utility function to extract a full uint256 from a field element and the
+    /// n-th set of residual bits from `residual`. This function is
+    /// curve-dependent.
+    function _extractBytes32(
```

Comment:
TODO: Add to the code conventions that virtual functions should be moved at the end of the file (it's more readable and easier to see which function must be implemented to have "concrete" contracts)